shivyc/tree/nodes.py

```diff
@@ -542,42 +544,56 @@ def get_storage(self, spec_kinds, spec_range):
 
         return storage
 
-    def parse_struct_spec(self, node, redec, symbol_table):
-        """Parse a struct ctype from the given decl_nodes.Struct node.
+    def parse_struct_union_spec(self, node, redec, symbol_table):
+        """Parse struct or union ctype from the given decl_nodes.Struct node.
 
-        node (decl_nodes.Struct) - the Struct node to parse
+        node (decl_nodes.Struct) - the Struct or Union node to parse
         redec (bool) - Whether this declaration is alone like so:
 
            struct S;
+           union U;
 
         or declares variables/has storage specifiers:
 
            struct S *p;
            extern struct S;
+           union U *u;
+           extern union U;
 
         If it's the first, then this is always a forward declaration for a
         new `struct S` but if it's the second and a `struct S` already
         exists in higher scope, it's just using the higher scope struct.
         """
         has_members = node.members is not None
+        kind = node.kind
         if node.tag:
             tag = str(node.tag)
-            ctype = symbol_table.lookup_struct(tag)
+            ctype = None
+            if kind == token_kinds.struct_kw:
+                ctype = symbol_table.lookup_struct(tag)
+            elif kind == token_kinds.union_kw:
+                ctype = symbol_table.lookup_union(tag)
 
             if not ctype or has_members or redec:
-                ctype = symbol_table.add_struct(tag, StructCType(tag))
+                if kind == token_kinds.struct_kw:
+                    ctype = symbol_table.add_struct_union(tag,
+                                                          StructCType(tag))
+                elif kind == token_kinds.union_kw:
+                    ctype = symbol_table.add_struct_union(tag,
+                                                          UnionCType(tag))
 
             if has_members and ctype.is_complete():
-                err = f"redefinition of 'struct {tag}'"
+                err = f"redefinition of '{tag}'"
```

Comment:
You can add back `struct` or `union` in this error message by using `{kind}` like you did below. (That was clever, by the way!)